### PR TITLE
QUIC: Raise more errors when appropriate

### PR DESCRIPTION
**Pull request by t8m**
There are probably more places where we could raise errors but this should be sufficient for MVP.

### CODE REVIEW

**Comment by mattcaswell** [comment]
CI failures are relevant

File: ssl/quic/quic_txp.c
Comment (h-vetinari) [inline ssl/quic/quic_txp.c:840]: Maybe move the comment up to 831 as well?

Comment (t8m) [reply ssl/quic/quic_txp.c:840]: I think the information from the comment is quite obvious now that we do not treat failures as OK if we sent at least one packet.

File: ssl/quic/quic_impl.c
Comment (mattcaswell) [inline ssl/quic/quic_impl.c:2710]: There is an inconsistency in this function now (but github won't let me comment on the right line).

Above we call `expect_quic_conn_only`, but if that fails it calls `QUIC_RAISE_NON_NORMAL_ERROR`. Although I'm not sure if that is correct - because that has the effect of causing `SSL_get_error()` to return `SSL_ERROR_SSL` which should only really happen for "IO" functions (which this isn't).

Below we just call `ERR_raise`, but not `QUIC_RAISE_NON_NORMAL_ERROR`.

File: ssl/quic/quic_impl.c
Comment (mattcaswell) [inline ssl/quic/quic_impl.c:3010]: Another case where this calls `QUIC_RAISE_NON_NORMAL_ERROR` but below we call ERR_raise. In this case this function is only ever called by the test framework though.

File: ssl/quic/quic_impl.c
Comment (mattcaswell) [inline ssl/quic/quic_impl.c:3087]: And again `expect_quic_conn_only` above can call `QUIC_RAISE_NON_NORMAL_ERROR`

File: ssl/quic/quic_record_shared.c
Comment (mattcaswell) [inline ssl/quic/quic_record_shared.c:150]: We're calling `tls13_hkdf_expand_ex` above with `raise_err` as 0 to suppress error raising...but below we're adding explicit error raising - which seems inconsistent. Also there are a number of cases above where we just "return 0" and don't  raise an error.

File: ssl/quic/quic_record_tx.c
Comment (mattcaswell) [inline ssl/quic/quic_record_tx.c:513]: Again there are cases above where we simply return 0 without raising an error, but here we are.

File: ssl/quic/quic_record_tx.c
Comment (mattcaswell) [inline ssl/quic/quic_record_tx.c:555]: Case below where we don't raise an error.

File: ssl/quic/quic_txp.c
Comment (mattcaswell) [inline ssl/quic/quic_txp.c:844]: This change is not obviously correct to me. Previously we returned `TX_PACKETISER_RES_SENT_PKT` as long as one packet was successfully committed. Now we return TX_PACKETISER_RES_FAILURE if any of them fail to commit.

File: ssl/quic/quic_wire_pkt.c
Comment (mattcaswell) [inline ssl/quic/quic_wire_pkt.c:44]: Again there are cases here where we raise an err, but in other codepaths above we do not.

File: ssl/quic/quic_wire_pkt.c
Comment (mattcaswell) [inline ssl/quic/quic_wire_pkt.c:89]: Err not raised on all codepaths.

File: ssl/quic/quic_wire_pkt.c
Comment (mattcaswell) [inline ssl/quic/quic_wire_pkt.c:868]: Err not raised on all codepaths.

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: ssl/quic/quic_txp.c
Comment (t8m) [reply ssl/quic/quic_txp.c:844]: I was contemplating on whether to make this change or not but IMO failing to commit a packet indicates a serious internal error and we should not continue. This allows indicating the internal error right away. @hlandau what's your opinion?

Comment (hlandau) [reply ssl/quic/quic_txp.c:844]: The original code was intentional. If we encounter an error of some kind after generating one or more packets successfully, reporting that we sent packets should take priority over reporting the error. A subsequent call can try again, fail for the same reason and report that error. So this change should be backed out.

File: ssl/quic/quic_channel.c
Comment (hlandau) [inline ssl/quic/quic_channel.c:2010]: Failure of `ch_retry` is not a protocol error. See the first comment in ch_retry &mdash; if a packet isn't handled we just drop it. As this is currently written a malicious third party could force connection termination by sending a packet with the wrong retry SCID...

File: ssl/quic/quic_impl.c
Comment (hlandau) [reply ssl/quic/quic_impl.c:1704]: These should 100% be using `QUIC_RAISE_NON_NORMAL_ERROR`. This macro should be used for anything dispatched via an API function (i.e., via `ssl_lib.c`) or where one of the `expect_quic` family of functions has been called on the call path. I'd be surprised if there are valid cases for direct calls to ERR_raise in `quic_impl.c` in particular. It would certainly be unusual.

File: ssl/quic/quic_impl.c
Comment (hlandau) [reply ssl/quic/quic_impl.c:2710]: If you don't want `last_error` changed for these calls, i would suggest the following changes:

- add a `is_io` argument to `quic_raise_non_normal_error` and make `last_error` setting conditional on this;
- add a new macro `QUIC_RAISE_NON_IO_ERROR` which is otherwise identical to `QUIC_RAISE_NON_NORMAL_ERROR`.

Error handling in the APL (`quic_impl.c`) should always be done in a principled fashion centralised through one of these macros.

**Review by hlandau — CHANGES_REQUESTED** [review]
Generally good stuff. @mattcaswell's comments apply. The TXP change needs to be reverted and all raising in `quic_impl.c` should use one of the RAISE macros, defining a new non-I/O one if desired.

File: ssl/quic/quic_txp.c
Comment (hlandau) [reply ssl/quic/quic_txp.c:844]: Why don't we change generate to use a separate `pkts_sent` output argument, which is always written on all escape paths from generate()?

Then the code under `TX_PACKETISER_RES_SENT_PKT` can be keyed on `if (pkts_sent > 0)` instead and the code under `default:` can be executed if the packetiser fails. Note that if it fails and has sent packets both of these blocks will execute, in that order.

Comment (t8m) [reply ssl/quic/quic_txp.c:844]: OK, I will do that

File: ssl/quic/quic_channel.c
Comment (t8m) [reply ssl/quic/quic_channel.c:2210]: @hlandau Really? The case you talk about returns 1 from ch_retry(). If an error is returned from ch_retry(), it is an internal error and IMO that should be handled.

File: ssl/quic/quic_record_tx.c
Comment (t8m) [reply ssl/quic/quic_record_tx.c:563]: ossl_quic_hdr_protector_encrypt() raises an error for us.

File: ssl/quic/quic_record_shared.c
Comment (t8m) [reply ssl/quic/quic_record_shared.c:156]: I've changed the tls13_hkdf_expand_ex() calls throughout the code to raise errors.

File: ssl/quic/quic_impl.c
Comment (t8m) [reply ssl/quic/quic_impl.c:2747]: Implemented as Hugo suggested.

File: ssl/quic/quic_impl.c
Comment (t8m) [reply ssl/quic/quic_impl.c:3049]: We now use QUIC_RAISE_NON_IO_ERROR in expect_quic_with_stream_lock() and below.

**Comment by t8m** [comment]
Ready for re-review.

File: ssl/quic/quic_channel.c
Comment (hlandau) [reply ssl/quic/quic_channel.c:2210]: See `ch_retry`:

```c
    /*
     * RFC 9000 s. 17.2.5.1: "A client MUST discard a Retry packet that contains
     * a SCID field that is identical to the DCID field of its initial packet."
     */
    if (ossl_quic_conn_id_eq(&ch->init_dcid, retry_scid))
        return 0;
```

You would at least need to change this to `return 1`. The other cases are probably OK to handle as internal errors.

Comment (t8m) [reply ssl/quic/quic_channel.c:2210]: Heh, I see return 1 in the branch there. I need to rebase it as it was changed later to return 0.

**Review by hlandau — APPROVED** [review]

**Comment by t8m** [comment]
Ping @openssl/committers for second review

**Review by mattcaswell — APPROVED** [review]

**Pull request closed without merging**
